### PR TITLE
Pubsub receipting

### PR DESCRIPTION
**Pull request by LukeLoze1**
# Motivation and Context
Brings receipting to the new case-processor

# What has changed
New listener for the Case.Responses Queue
Emits UAC Updated event for the 1st uacQidLink for the passed case (until we get qid rather than case id)

# How to test?
When available the AT's should exercise the code,
will need to have:
 https://github.com/ONSdigital/census-rm-pubsub metadata-json branch
&
https://github.com/ONSdigital/census-rm-acceptance-tests/branches : pubsub-receipting

# Links
https://trello.com/c/W5MYJG94/654-rmc-177-manage-uac-qid-pairs-for-welsh-initial-contact-questionnaire-print-packs-13
https://github.com/ONSdigital/census-rm-pubsub/pull/21

# Screenshots (if appropriate):

### CODE REVIEW

**Comment by LukeLoze1** [comment]
[Case processor should handle receipt message from pubsub service (8)](https://trello.com/c/2lcQkGCO/674-case-processor-should-handle-receipt-message-from-pubsub-service-8)

**Comment by LukeLoze1** [comment]
[Case processor should handle receipt message from pubsub service (8)](https://trello.com/c/2lcQkGCO/674-case-processor-should-handle-receipt-message-from-pubsub-service-8)

**Comment by LukeLoze1** [comment]
[Case processor should handle receipt message from pubsub service (8)](https://trello.com/c/2lcQkGCO/674-case-processor-should-handle-receipt-message-from-pubsub-service-8)

File: pom.xml [hunk 1]
```diff
@@ -75,6 +75,11 @@
       <artifactId>spring-boot-starter-test</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.jeasy</groupId>
+      <artifactId>easy-random-core</artifactId>
+      <version>4.0.0.RC1</version>
```
Comment (NickGrantONS) [inline pom.xml]: this should include `<scope>test</scope>`

File: src/main/java/uk/gov/ons/census/casesvc/model/repository/CaseRepository.java [hunk 2]
```diff
@@ -1,7 +1,10 @@
 package uk.gov.ons.census.casesvc.model.repository;
 
+import java.util.Optional;
 import java.util.UUID;
 import org.springframework.data.jpa.repository.JpaRepository;
 import uk.gov.ons.census.casesvc.model.entity.Case;
 
-public interface CaseRepository extends JpaRepository<Case, UUID> {}
+public interface CaseRepository extends JpaRepository<Case, UUID> {
+  Optional<Case> findByCaseId(UUID caseId);
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/casesvc/model/repository/CaseRepository.java]: is it really necessary to wrap the result in an `Optional`?

File: src/main/java/uk/gov/ons/census/casesvc/service/EventProcessor.java [hunk 3]
```diff
@@ -28,15 +24,15 @@ public void processSampleReceivedMessage(CreateCaseSample createCaseSample) {
     int questionnaireType =
         QuestionnaireTypeHelper.calculateQuestionnaireType(caze.getTreatmentCode());
     UacQidLink uacQidLink = uacProcessor.saveUacQidLink(caze, questionnaireType);
-    uacProcessor.emitUacUpdatedEvent(uacQidLink, caze);
+    uacProcessor.emitUacUpdatedEvent(uacQidLink, caze, true);
     caseProcessor.emitCaseCreatedEvent(caze);
-    uacProcessor.logEvent(uacQidLink, CASE_CREATED_EVENT_DESCRIPTION);
-    uacProcessor.logEvent(uacQidLink, UAC_QID_LINKED_EVENT_DESCRIPTION);
+    uacProcessor.logEvent(uacQidLink, CASE_CREATED_EVENT_DESCRIPTION, null);
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/casesvc/service/EventProcessor.java]: can't you overload the `logEvent` method so that it only takes 2 parameters and defaults the third to null?

File: src/main/java/uk/gov/ons/census/casesvc/service/UacProcessor.java [hunk 4]
```diff
@@ -57,24 +57,27 @@ public UacQidLink saveUacQidLink(Case caze, int questionnaireType) {
     return uacQidLink;
   }
 
-  public void logEvent(UacQidLink uacQidLink, String eventDescription) {
+  public void logEvent(
+      UacQidLink uacQidLink, String eventDescription, LocalDateTime eventMetaDataDateTime) {
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/casesvc/service/UacProcessor.java]: suggestion: overload this method so that this method gets called with `eventMetaDataDateTime` as null if the parameter is not specified

File: src/main/java/uk/gov/ons/census/casesvc/service/UacProcessor.java [hunk 5]
```diff
@@ -57,24 +57,27 @@ public UacQidLink saveUacQidLink(Case caze, int questionnaireType) {
...
-    loggedEvent.setEventDate(new Date());
+    loggedEvent.setEventDate(eventMetaDataDateTime);
+    loggedEvent.setRmEventProcessed(LocalDateTime.now());
     loggedEvent.setEventDescription(eventDescription);
     loggedEvent.setUacQidLink(uacQidLink);
     eventRepository.save(loggedEvent);
   }
 
-  public void emitUacUpdatedEvent(UacQidLink uacQidLink, Case caze) {
+  public void emitUacUpdatedEvent(UacQidLink uacQidLink, Case caze, Boolean active) {
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/casesvc/service/UacProcessor.java]: 1. I hate the use of `Boolean` except where were' planning on passing a null as one of three options

2. Can't we overload this method so that we keep the original 2-argument version, calling the second one with null or whatever default we want?

File: src/main/java/uk/gov/ons/census/casesvc/utility/Sha256Helper.java [hunk 6]
```diff
@@ -6,7 +6,7 @@
 import javax.xml.bind.DatatypeConverter;
 
 public class Sha256Helper {
-  private static MessageDigest digest;
+  private static final MessageDigest digest;
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/casesvc/utility/Sha256Helper.java]: I feel so much safer now

File: src/main/java/uk/gov/ons/census/casesvc/model/dto/Receipt.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package uk.gov.ons.census.casesvc.model.dto;
+
+import java.time.LocalDateTime;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+public class Receipt {
+  private String case_id;
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/casesvc/model/dto/Receipt.java]: Not camel case. Can we not use some kind of `@Json` Jackson unmarshalling annotation to get around this horribleness?

Comment (NickGrantONS) [reply src/main/java/uk/gov/ons/census/casesvc/model/dto/Receipt.java]: use `@JsonProperty("case_id")` instead of breaking code style rules

File: src/main/java/uk/gov/ons/census/casesvc/model/dto/Receipt.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  private String tx_id;
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/casesvc/model/dto/Receipt.java]: use @JsonProperty("tx_id") instead of breaking code style rules

File: src/main/java/uk/gov/ons/census/casesvc/model/dto/Receipt.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+  private String questionnaire_id;
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/casesvc/model/dto/Receipt.java]: use @JsonProperty("questionnaire_id") instead of breaking code style rules

File: src/main/java/uk/gov/ons/census/casesvc/model/dto/Receipt.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+  private LocalDateTime response_dateTime;
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/casesvc/model/dto/Receipt.java]: use @JsonProperty("response_dateTime") instead of breaking code style rules

File: src/main/java/uk/gov/ons/census/casesvc/model/dto/Receipt.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+  private String inbound_channel;
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/casesvc/model/dto/Receipt.java]: use @JsonProperty("inbound_channel") instead of breaking code style rules

File: src/test/java/uk/gov/ons/census/casesvc/testutil/RabbitQueueHelper.java [hunk 12]
```diff
@@ -24,7 +24,7 @@
 @ActiveProfiles("test")
 @EnableRetry
 public class RabbitQueueHelper {
-  private ObjectMapper objectMapper = new ObjectMapper();
+  private final ObjectMapper objectMapper = new ObjectMapper();
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/census/casesvc/testutil/RabbitQueueHelper.java]: disaster averted

File: src/main/java/uk/gov/ons/census/casesvc/model/repository/CaseRepository.java (hunk 2 above)
Comment (LukeLoze1) [reply src/main/java/uk/gov/ons/census/casesvc/model/repository/CaseRepository.java]: My understanding is that since Java 8 you should wrap a potential Null in an optional.  This is used with 'free' JPA functions like findById, so also feels more consistent.

File: .travis.yml [hunk 13]
```diff
@@ -30,3 +30,4 @@ cache:
 branches:
   only:
   - master
+  - pubsub-receipting
```
Comment (NickGrantONS) [inline .travis.yml]: why is this needed?

File: src/main/java/uk/gov/ons/census/casesvc/model/dto/Receipt.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+  @JsonProperty("case_id")
+  private String caseId;
+
+  @JsonProperty("tx_id")
+  private String txId;
+
+  @JsonProperty("questionnaire_id")
+  private String questionnaire_Id;
+
+  @JsonProperty("response_dateTime")
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/casesvc/model/dto/Receipt.java]: the little-known camelSnake_case lol

File: src/main/java/uk/gov/ons/census/casesvc/service/ReceiptProcessor.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    this.caseRepository = caseRepository;
+    this.uacProcessor = uacProcessor;
+  }
+
+  public void processReceipt(Receipt receipt) {
+    // HERE BE DRAGONS, THIS IS A HACK.  IN THE LONG RUN WE WILL RECEIVE JUST A QID
+    // HOWEVER THIS CODE IS WRITTEN IN A WAY TO MAKE THE PROMISED LAND OF RECEIVING A QID EASY
+    // JUST HAVE A QIDREPOSITORY RATHER THAN A CASE RESPOSITORY AND WORK OF THAT (AND THE QID)
+
+    Optional<Case> cazeOpt = caseRepository.findByCaseId(UUID.fromString(receipt.getCaseId()));
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/casesvc/service/ReceiptProcessor.java]: don't we want to mark on the case that we've received a response, or is that out of scope for this ticket?

**Comment by jcox-dev** [comment]
Closing in favour of #12
